src/main/java/jnr/ffi/Platform.java

```diff
@@ -474,43 +467,91 @@ public boolean accept(File dir, String name) {
                 exclude = Pattern.compile(".*(lib[a-z]*64|amd64|x86_64).*");
             }
 
-            List<File> matches = new LinkedList<File>();
-            for (String path : libraryPath) {
+            final Pattern versionedLibPattern = Pattern.compile("lib" + libName + "\\.so((?:\\.[0-9]+)*)$");
+
+            FilenameFilter filter = new FilenameFilter() {
+                public boolean accept(File dir, String name) {
+                    return versionedLibPattern.matcher(name).matches();
+                }
+            };
+
+            Map<String, int[]> matches = new HashMap<String, int[]>();
+            for (String path : libraryPaths) {
                 if (exclude.matcher(path).matches()) {
                     continue;
                 }
-                File[] files = new File(path).listFiles(filter);
-                if (files != null && files.length > 0) {
-                    matches.addAll(Arrays.asList(files));
+
+                File libraryPath = new File(path);
+                File[] files = libraryPath.listFiles(filter);
+                if (files == null) {
+                    continue;
+                }
+
+                for (File file : files) {
+                    Matcher matcher = versionedLibPattern.matcher(file.getName());
+                    String versionString = matcher.matches() ? matcher.group(1) : "";
+                    int[] version;
+                    if (versionString == null || versionString.isEmpty()) {
+                        version = new int[0];
+                    } else {
+                        String[] parts = versionString.split("\\.");
+                        version = new int[parts.length - 1];
+                        for (int i = 1; i < parts.length; i++) {
+                            version[i - 1] = Integer.parseInt(parts[i]);
+                        }
+                    }
+                    matches.put(file.getAbsolutePath(), version);
```

Comment:
This loses the ordering of the library paths and is proving to be problematic for me. I have both `/lib64/librt.so.1` and '/lib/librt.so.1`, with `/lib64` before `/lib`. I'm going to look at stripping out the `/lib` part, but it's coming from the "java.library.path" system property so it's likely to be there for others to. In any event, since the HashMap here loses the ordering, `/lib/librt.so1` actually comes before `/lib64/librt.so.1` during the best match scan. Since they have the same version number, there's no way for the `/lib64` version to win and I wind up with an `UnsatisfiedLinkError: /lib/librt.so.1: wrong ELF class: ELFCLASS32`.